README.md
<p align="center">
<a title="Pub" href="https://pub.dartlang.org/packages/flame_composition_rendering" ><img alt="Pub Version" src="https://img.shields.io/pub/v/flame_composition_rendering?color=blue&style=for-the-badge"></a>
<a title="Pub" href="https://pub.dartlang.org/packages/composition_rendering" ><img alt="Pub Version" src="https://img.shields.io/pub/v/composition_rendering?label=composition_rendering&style=for-the-badge"></a>
</p>

## Usage FlamePluginComponent

A simple usage example:

```dart
import 'package:composition_rendering/scene.dart';
import 'package:flame/flame.dart';
import 'package:flame/game.dart';
import 'package:flame_composition_rendering/flame_plugin.dart';
import 'package:flutter/widgets.dart';

void main() {
  WidgetsFlutterBinding.ensureInitialized();

  runApp(
    GameWidget(
      game: ExampleLauncher(),
    ),
  );
}

class ExampleLauncher extends BaseGame {
  final FlamePluginComponent plugin = FlamePluginComponent();

  ExampleLauncher() {
    plugin.scene = GameScene();
  }

  @override
  Future<void> onLoad() async {
    await Flame.images.loadAll([
      'bg.jpg',
    ]);
    await add(plugin);
  }
}

class GameScene extends Scene {
  @override
  void init() {
    // your play scene
  }
}
```


## Usage SpriteBatchComponent

Alternative SpriteBatch for Flame. Rendering three sprites in one call [Canvas.drawRawAtlas](https://api.flutter.dev/flutter/dart-ui/Canvas/drawRawAtlas.html)

```dart
import 'package:flame/components.dart';
import 'package:flame/flame.dart';
import 'package:flame/game.dart';
import 'package:flame_composition_rendering/flame_composition_rendering.dart';
import 'package:flutter/material.dart';

void main() {
  WidgetsFlutterBinding.ensureInitialized();

  Flame.device.fullScreen();
  Flame.device.setLandscape();

  runApp(
    GameWidget(
      game: ExampleLauncher(),
    ),
  );
}

class ExampleLauncher extends BaseGame {
  @override
  Future<void> onLoad() async {
    await Flame.images.loadAll(['crab.png']);
    await add(SpriteBatchComponent());
  }

  @override
  Color backgroundColor() => const Color(0xFFFFFFFF);
}

class SpriteBatchComponent extends FlameBatchComponent {
  late final image;
  late final sprite;

  @override
  Future<void>? onLoad() {
    image = Flame.images.fromCache('crab.png');
    sprite = Sprite(image);
  }

  @override
  void beginBatch() {
    drawSprite(sprite, x: 100, y: 200);
    drawSprite(sprite, x: 200, y: 200);
    drawSprite(sprite, x: 300, y: 200);
    drawSpriteFromComponents(sprite, x: 400, y: 100, rotation: 0.4, scale: 0.6);
  }
}
```

CHANGELOG.md
## 0.1.0-dev.7 (2021-10-24)

* Update composition_rendering: ^0.4.0
* Update flame: ^1.0.0-releasecandidate.16

## 0.1.0-dev.6 (2021-10-05)

* Update composition_rendering: ^0.3.1
* Update flame: ^1.0.0-releasecandidate.15

## 0.1.0-dev.5 (2021-09-23)

* Update composition_rendering: ^0.3.0

## 0.1.0-dev.4

* Update flame: ^1.0.0-releasecandidate.14

## 0.1.0-dev.3

* Update composition_rendering: ^0.2.1

## 0.1.0-dev.2

* Update composition_rendering: ^0.2.0

## 0.1.0-dev.1

* Initial version, created by QiXi

example/flame_plugin_example.dart
import 'package:composition_rendering/scene.dart';
import 'package:flame/flame.dart';
import 'package:flame/game.dart';
import 'package:flame_composition_rendering/flame_composition_rendering.dart';
import 'package:flutter/widgets.dart';

void main() {
  WidgetsFlutterBinding.ensureInitialized();

  Flame.device.fullScreen();
  Flame.device.setLandscape();
  runApp(
    GameWidget(
      game: ExampleLauncher(),
    ),
  );
}

class ExampleLauncher extends FlameGame {
  final FlamePluginComponent plugin = FlamePluginComponent();
  final GameScene gameScene = GameScene();

  ExampleLauncher() {
    plugin.scene = gameScene;
  }

  @override
  Future<void> onLoad() async {
    await Flame.images.loadAll([
      'bg.jpg',
    ]);
    await add(plugin);
  }
}

class GameScene extends Scene {
  @override
  void init() {
    // your play scene
    //add(sprite)
  }
}


lib/src/flame_batch_component.dart
import 'dart:ui';

import 'package:composition_rendering/batch.dart';
import 'package:flame/components.dart';

abstract class FlameBatchComponent extends Component {
  EngineBatch batch;
  Rect? cullRect;

  FlameBatchComponent() : batch = EngineBatch();

  FlameBatchComponent.fromEngineBatch(
    this.batch, {
    this.cullRect,
  });

  void beginBatch();

  void drawSprite(Sprite sprite, {double x = 0, double y = 0}) {
    batch.drawImageFromData(sprite.image, sprite.src, 1, 0, x, y, sprite.paint.color.value);
  }

  void drawSpriteFromComponents(
    Sprite sprite, {
    double rotation = 0,
    double scale = 1,
    double? anchorX,
    double? anchorY,
    double x = 0,
    double y = 0,
  }) {
    batch.drawImageFromComponents(
        image: sprite.image,
        rect: sprite.src,
        rotation: rotation,
        scale: scale,
        anchorX: anchorX ?? -sprite.src.left + sprite.src.width / 2.0,
        anchorY: anchorY ?? -sprite.src.top + sprite.src.height / 2.0,
        translateX: x,
        translateY: y);
  }

  @override
  void render(Canvas canvas) {
    batch.beginBatch(canvas, cullRect: cullRect);
    beginBatch();
    batch.endBatch();
  }
}


lib/src/flame_plugin_component.dart
import 'dart:ui' show Canvas;

import 'package:composition_rendering/core.dart';
import 'package:composition_rendering/scene.dart';
import 'package:composition_rendering/systems.dart';
import 'package:flame/components.dart' show Component;

import 'plugin_system_registry.dart';

class FlamePluginComponent extends Component with Registry {
  static final PluginSystemRegistry registry = PluginSystemRegistry();

  late final Scene _scene;

  FlamePluginComponent() {
    SystemRegistry.instance = registry;
    changePriorityWithoutResorting(50);
  }

  set scene(Scene scene) {
    _scene = scene;
  }

  @override
  Future<void>? onLoad() async {
    _scene.init();
  }

  void onResize(Vector2 canvasSize) {
    parameters.setViewSize(canvasSize.x, canvasSize.y);
  }

  @override
  void update(double dt) {
    super.update(dt);
    systems.updateSystem.update(dt);
    _scene.updateScene(dt);
  }

  @override
  void render(Canvas canvas) {
    systems.renderSystem.render(canvas);
  }
}


example/flame_batch_example.dart
import 'package:flame/components.dart';
import 'package:flame/flame.dart';
import 'package:flame/game.dart';
import 'package:flame_composition_rendering/flame_composition_rendering.dart';
import 'package:flutter/material.dart';

void main() {
  WidgetsFlutterBinding.ensureInitialized();

  Flame.device.fullScreen();
  Flame.device.setLandscape();

  runApp(
    GameWidget(
      game: ExampleLauncher(),
    ),
  );
}

class ExampleLauncher extends FlameGame {
  @override
  Future<void> onLoad() async {
    await Flame.images.loadAll(['crab.png']);
    await add(SpriteBatchComponent());
  }

  @override
  Color backgroundColor() => const Color(0xFFFFFFFF);
}

class SpriteBatchComponent extends FlameBatchComponent {
  late final image;
  late final sprite;

  @override
  Future<void>? onLoad() {
    image = Flame.images.fromCache('crab.png');
    sprite = Sprite(image);
  }

  @override
  void beginBatch() {
    drawSprite(sprite, x: 100, y: 200);
    drawSprite(sprite, x: 200, y: 200);
    drawSprite(sprite, x: 300, y: 200);
    drawSpriteFromComponents(sprite, x: 400, y: 100, rotation: 0.4, scale: 0.6);
  }
}
